Include/internal/pycore_code.h

```diff
@@ -24,6 +26,85 @@ struct _PyOpcache {
     char optimized;
 };
 
+
+/* PEP 659
+ * Specialization and quickening structs and helper functions
+ */
+
+typedef struct {
+    int32_t cache_count;
+    int32_t _;
+} EntryZero;
+
+/* Add specialized versions of entries to this union.
+ * Do not break this invariant: sizeof(SpecializedCacheEntry) == 8 */
+typedef union {
+    EntryZero zero;
+    PyObject *object;
+} SpecializedCacheEntry;
+
+#define INSTRUCTIONS_PER_ENTRY (sizeof(SpecializedCacheEntry)/sizeof(_Py_CODEUNIT))
+
+/* Maximum size of code to quicken */
+#define MAX_SIZE_TO_QUICKEN 5000
+
+typedef union _cache_or_instruction {
+    _Py_CODEUNIT code[1];
+    SpecializedCacheEntry entry;
+} SpecializedCacheOrInstruction;
+
+static inline SpecializedCacheEntry *
+_GetSpecializedCacheEntry(_Py_CODEUNIT *first_instr, Py_ssize_t index)
+{
+    SpecializedCacheOrInstruction *last_cache_plus_one = (SpecializedCacheOrInstruction *)first_instr;
+    assert(&last_cache_plus_one->code[0] == first_instr);
+    return &last_cache_plus_one[-1-index].entry;
+}
+
+/* Following two functions determine the index of a cache entry from the
+ * instruction index (in the instruction array) and the oparg.
+ * oparg_from_offset_and_index must be the inverse of
+ * offset_from_oparg_and_index
+ */
+
+static inline int
+oparg_from_offset_and_index(int offset, int index)
+{
+    return offset-(index>>1);
+}
+
+static inline int
+offset_from_oparg_and_index(int oparg, int index)
+{
+    return (index>>1)+oparg;
+}
+
+static inline SpecializedCacheEntry *
+_GetSpecializedCacheEntryForInstruction(_Py_CODEUNIT *first_instr, int index, int oparg)
+{
+    return _GetSpecializedCacheEntry(
+        first_instr,
+        offset_from_oparg_and_index(oparg, index)
+    );
+}
+
+#define QUICKENING_INITIAL_WARMUP_DELAY 8
```

Comment:
Add comments explaining the meaning of these constants? At least link to co_warmup. I was confused because the warmup delay is negated before setting co_warmup, while the coldest value is not. If the counter counts up from -8, maybe make the sign part of the constant?